pisa/utils/hyperplane.py

```diff
@@ -0,0 +1,628 @@
+                return self.evaluate(sys_params_unflattened,bin_idx=bin_idx)
+
+
+            #
+            # Fit
+            #
+
+            # Define the EPS (step length) used by the fitter
+            # Need to take care with floating type precision, don't want to go smaller than the FTYPE being used by PISA can handle
+            eps = np.finfo(FTYPE).eps
+ 
+            # Perform fit
+            #TODO limit all params to [0,1] as we do for minimizers?
+            popt, pcov = curve_fit(
+                callback,
+                x,
+                y,
+                p0=p0,
+                maxfev=1000,
+                epsfcn=eps,
+            )
+
+            # Check the fit was successful
+            #TODO
+
+            # Write the fit results back to the hyperplane structure
+            self._intercept[bin_idx] = popt[0]
+            i = 1
+            for sys_param in self._sys_params.values() :
+                for j in range(sys_param._num_fit_params) :
+                    sys_param._fit_params[bin_idx,j] = popt[i]
+                    i += 1
+
+        #
+        # Done
+        #
+
+        # Record some provenance info about the fits
+        self._fit_complete = True
+        self._fit_normed_maps = norm
+        self._fit_maps_raw = maps
+        self._fit_maps_used = maps_to_use
+
+
+    def get_on_axis_mask(self,sys_param_name) :
+        '''
+        TODO
+        '''
+
+        assert sys_param_name in self.sys_param_names
+
+        num_fitting_datasets = self._sys_params.values()[0]._fitting_sys_dataset_values.size
+        on_axis_mask = np.ones((num_fitting_datasets,),dtype=bool)
+
+        # Loop over sys params
+        for sys_param in self._sys_params.values() :
+
+            # Ignore the chosen param
+            if sys_param.name  != sys_param_name :
+
+                # Define a "nominal" mask
+                on_axis_mask = on_axis_mask & np.isclose(sys_param._fitting_sys_dataset_values,sys_param.nominal_value) 
+
+        return on_axis_mask
+
+
+    def report(self,bin_idx=None) :
+        '''
+        String version of the hyperplane contents
+        '''
+
+        # Fit results
+        print(">>>>>> Fit parameters >>>>>>")
+        bin_indices = np.ndindex(self._binning.shape) if bin_idx is None else [bin_idx]
+        for bin_idx in bin_indices :
+            print("  Bin %s :" % (bin_idx,) )
+            print("     Intercept : %0.3g" % (self._intercept[bin_idx],) )
+            for sys_param in self._sys_params.values() :
+                print("     %s : %s" % ( sys_param.name, ", ".join([ "%0.3g"%sys_param._fit_params[(bin_idx,i,)] for i in range(sys_param._num_fit_params) ])) )
+        print("<<<<<< Fit parameters <<<<<<")
+
+
+
+class SysParam(object) :
+    '''
+    A class defining the systematic parameter in the hyperplane
+    Use constructs this by passing the functional form (as a function)
+    '''
+
+    def __init__(self,name,nominal_value,func_name,initial_fit_params=None) :
```

Comment:
can you make `initial_fit_params` a dict instead of a list so the user explicitly sets param values by name, or does that make things tricky / worse elsewhere? I always prefer named things, but if it makes other code difficult/impossible/ugly, then I understand making these a simple list of values.